ambry-store/src/main/java/com/github/ambry/store/ReplicaPlacementValidator.java

```diff
@@ -0,0 +1,146 @@
+/**
+ * Copyright 2024 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.store;
+
+import com.github.ambry.clustermap.DiskId;
+import com.github.ambry.clustermap.ReplicaId;
+import java.io.File;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+
+/**
+ * Loop through all the host's disks and check if the replicas are placed correctly
+ * (as compared with the latest Helix state). Sometimes, disks can get re-mounted in the wrong
+ * order. This class checks for such cases and reshuffles the replicas if necessary, creating
+ * a map of the broken disks and the disks they should be reshuffled to.
+ */
+public class ReplicaPlacementValidator {
+  private final Map<DiskId, Set<String>> foundDiskToPartitionMap;
+  private Map<DiskId, List<ReplicaId>> expectedDiskToReplicaMap;
+  private Map<DiskId, List<ReplicaId>> newDiskToReplicaMap;
+  private List<DiskId> brokenDisks = new ArrayList<>();
+  public ReplicaPlacementValidator(Map<DiskId, List<ReplicaId>> expectedDiskToReplicaMap) {
+    foundDiskToPartitionMap = new HashMap<>();
+    newDiskToReplicaMap = new HashMap<>();
+    this.expectedDiskToReplicaMap = expectedDiskToReplicaMap;
+
+    for (Map.Entry<DiskId, List<ReplicaId>> entry : expectedDiskToReplicaMap.entrySet()) {
+      DiskId currentDisk = entry.getKey();
+      foundDiskToPartitionMap.put(currentDisk, findPartitionsOnDisk(currentDisk));
+      List<ReplicaId> replicas = entry.getValue();
+      for (ReplicaId replica : replicas) {
+        String partitionID = replica.getPartitionId().toString();
+        if(!foundDiskToPartitionMap.get(currentDisk).contains(partitionID)) {
+          brokenDisks.add(currentDisk);
+        }
```

Comment:
If a host got reimaged, and came back with empty disks, then we don't want to think those disks are broken.

Reply:
Yeah, this is intended. As discussed offline, the code in this pull request just attempts to rectify the simplest possible case - and by design does not attempt to do anything if any disks are empty. We'll revisit this design decision later if it turns out to be inadequate in practice.